service/tbc/tbc.go

```diff
@@ -1015,96 +993,80 @@ func (s *Server) downloadBlock(ctx context.Context, p *peer, ch *chainhash.Hash)
+	log.Tracef("blockExpired")
+	defer log.Tracef("blockExpired exit")
 
-	now := time.Now()
+	p, ok := value.(*peer)
+	if !ok {
+		// this really should not happen
+		log.Errorf("block expired no peer: %v", key)
+		return
+	}
+	log.Infof("block expired %v: %v", p, key)
 
-	defer func() {
-		// if we are complete we need to kick off utxo sync
-		go s.utxoIndexer(ctx)
-	}()
+	p.close() // this will tear down peer
 
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
+	// XXX we really should scan all pending blocks here to remove all
+	// pending block on this peer. This does correct itself so may not be
+	// worth the trouble.
+}
 
-	for k, v := range s.blocks {
-		if v != nil && now.After(v.expire) {
-			// kill peer as well since it is slow
-			if p := s.peers[v.peer]; p != nil && p.conn != nil {
-				p.conn.Close() // this will tear down peer
-			}
+// randomPeer returns a random peer from the map. Must be called with lock
+// held.
+func (s *Server) randomPeer(ctx context.Context) (*peer, error) {
+	log.Tracef("randomPeer")
+	defer log.Tracef("randomPeer exit")
 
-			// block expired, download block
-			s.blocks[k] = nil
-			v = nil // this will redownload the block
-		}
-		if v != nil {
-			// block already being downloaded and is not expired
+	// unassigned slot, download block
+	for _, p := range s.peers {
+		if p.conn == nil {
+			// Not connected yet
 			continue
 		}
-		// unassigned slot, download block
-		for _, peer := range s.peers {
-			if peer.conn == nil {
-				// Not connected yet
-				continue
-			}
-			ch, err := chainhash.NewHashFromStr(k)
-			if err != nil {
-				// really should not happen
-				log.Errorf("download blocks hash: %v", err)
-				delete(s.blocks, k)
-				continue
-			}
-
-			// sufficiently validated, record in cache
-			s.blocks[k] = &blockPeer{
-				expire: time.Now().Add(37 * time.Second), // XXX make variable?
-				peer:   peer.String(),
-			}
-
-			go s.downloadBlock(ctx, peer, ch)
-
-			break
-		}
+		return p, nil
 	}
+	return nil, fmt.Errorf("no peers")
 }
 
 func (s *Server) syncBlocks(ctx context.Context) {
 	log.Tracef("syncBlocks")
 	defer log.Tracef("syncBlocks exit")
 
-	// regardless of cache being full or no more missing blocks kick the
-	// downloader just to make sure we are making forward progress.
-	defer func() {
-		go s.downloadBlocks(ctx)
-	}()
-
-	// Hold lock to fill blocks cache
+	// Don't let want race. If we do the cache capacity will be overshot.
 	s.mtx.Lock()
 	defer s.mtx.Unlock()
 
-	// Deal with expired block downloads
-	used := s.blockPeerExpire()
-	want := defaultPendingBlocks - used
+	want := defaultPendingBlocks - s.blocks.Len()
 	if want <= 0 {
 		return
 	}
-
 	bm, err := s.db.BlocksMissing(ctx, want)
 	if err != nil {
 		log.Errorf("blocks missing: %v", err)
 		return
 	}
+
 	for k := range bm {
 		bi := bm[k]
 		hash, _ := chainhash.NewHash(bi.Hash[:])
 		hashS := hash.String()
-		if _, ok := s.blocks[hashS]; ok {
+		if _, _, err := s.blocks.Get(hashS); err == nil {
```

Comment:
Would it be worth adding a `Contains(key)` function to the TTL map? :thinking:

Reply:
I wasn't sold on having it just yet.